### PR TITLE
Drop older slots in the ledger

### LINKED ISSUE #5051: Facility to a validator to prune older parts of the ledger
#### Problem
Validators that don't care to retain the full ledger have no mechanism to manage their disk space beyond periodically stopping, deleting the ledger, and restarting from a snapshot.

#### Proposed Solution
Actively prune older slots out of blocktree, perhaps around here: https://github.com/solana-labs/solana/blob/9f58318fc59da276bc8b4014760032319b2263b6/core/src/bank_forks.rs#L138
The trick is that once the validator starts discarding old ledger, it can no longer participate in repairing other nodes that are behind the discard point.   So we'd want to keep some amount of ledger (maybe an hour or so of data at least) around to help out the nodes that are trying to catch up from a recent snapshot

### PR DESCRIPTION
**Pull request by sagar-solana**
#### Problem

Need a way to limit ledger disk space usage 

#### Summary of Changes

Actively prune older slots out of blocktree.

- Added a ledger cleanup service that runs off of replay stage signals
- Added tests (local_cluster test that runs with max_ledger_slots == 100 works!)
- Feature is optional and not enabled by default _and_ requires snapshots to be enabled

Fixes #5051

### CODE REVIEW

**Comment by sagar-solana** [comment]
@rob-solana @mvines @carllin, thoughts? I have not tried restarting a node that runs with this flag enabled...

File: core/src/blocktree.rs [hunk 1]
```diff
@@ -193,66 +193,69 @@ impl Blocktree {
         false
     }
 
-    // silently deletes all blocktree column families starting at the given slot
-    fn delete_all_columns(&self, starting_slot: u64) {
-        match self.meta_cf.force_delete_all(Some(starting_slot)) {
+    /// Silently deletes all blocktree column families starting at the given slot until the `to` slot
+    /// Use with care; does not check for integrity and does not update slot metas that
+    /// refer to deleted slots
+    pub fn delete_all_columns(&self, from_slot: u64, to_slot: Option<u64>) {
```
Comment (rob-solana) [inline core/src/blocktree.rs]: use Slot from the SDK for slots

Comment (rob-solana) [inline core/src/blocktree.rs]: can you pass an std::ops::Range to this function?

File: core/src/blocktree.rs [hunk 2]
```diff
@@ -193,66 +193,69 @@ impl Blocktree {
...
+        let to_index = to_slot.map(|slot| (slot + 1, 0));
+        match self.meta_cf.force_delete(Some(from_slot), to_slot) {
             Ok(_) => (),
             Err(e) => error!(
                 "Error: {:?} while deleting meta_cf for slot {:?}",
-                e, starting_slot
+                e, from_slot
```
Comment (rob-solana) [inline core/src/blocktree.rs]: this form might collapse some with map_err()

File: core/src/blocktree/db.rs [hunk 3]
```diff
@@ -409,13 +409,22 @@ where
...
+        C::Index: PartialOrd,
+    {
+        let iter = self.iter(from)?;
+        for (index, _) in iter {
+            if let Some(ref to) = to {
+                if &index > to {
+                    break;
+                }
+            }
+            match self.delete(index) {
```
Comment (rob-solana) [inline core/src/blocktree/db.rs]: another place for map_err()?

File: core/src/blocktree.rs (hunk 1 above)
Comment (sagar-solana) [reply core/src/blocktree.rs]: I tried using rangebounds (so the we can do `(n.. m)`) index doesn't support a default (to start at 0)and it basically became really hacky. A std::ops::Range _requires_ a lower and upper bound.

I can give it another shot but it requires index to turn into a more complex type that every cf will have to implement.

File: core/src/ledger_cleanup_service.rs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+use std::time::Duration;
+
+pub const DEFAULT_MAX_LEDGER_SLOTS: u64 = DEFAULT_SLOTS_PER_EPOCH * 5;
+
+pub struct LedgerCleanupService {
+    t_cleanup: JoinHandle<()>,
+}
+
+impl LedgerCleanupService {
+    #[allow(clippy::new_ret_no_self)]
```
Comment (rob-solana) [inline core/src/ledger_cleanup_service.rs]: is the clippy warning necessary?

File: core/src/blocktree.rs (hunk 2 above)
Comment (sagar-solana) [reply core/src/blocktree.rs]: I think clippy will complain if I map_err and just print, and map it to itself.

File: core/src/ledger_cleanup_service.rs (hunk 4 above)
Comment (sagar-solana) [reply core/src/ledger_cleanup_service.rs]: Yeah this seems unnecessary I must have been returning other crap earlier.

File: core/src/blocktree.rs (hunk 2 above)
Comment (sagar-solana) [reply core/src/blocktree.rs]: I could `if let Err(e)`...

File: core/src/ledger_cleanup_service.rs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+    fn cleanup_ledger(
+        slot_full_receiver: &Receiver<(u64, Pubkey)>,
+        blocktree: &Arc<Blocktree>,
+        max_ledger_slots: u64,
+    ) -> Result<()> {
+        let (slot, _) = slot_full_receiver.recv_timeout(Duration::from_secs(1))?;
+        if slot > max_ledger_slots {
+            //cleanup
+            blocktree.delete_all_columns(0, Some(slot - max_ledger_slots));
```
Comment (rob-solana) [inline core/src/ledger_cleanup_service.rs]: delete_all_columns() is a wrong name, conveys info about current implementation that might not survive a rewrite.

prune() isn't a wrong name, but is confusing if it takes 2 parameters.

how about transplant(), graft(), forget(), stipulate(), forswear(), forgo()? each of which would take a single parameter.

p.s. this looks scary

File: core/src/blocktree.rs (hunk 1 above)
Comment (rob-solana) [reply core/src/blocktree.rs]: not worth it

File: core/src/ledger_cleanup_service.rs (hunk 5 above)
Comment (sagar-solana) [reply core/src/ledger_cleanup_service.rs]: Updated it to `purge_slots`. Not terribly attached to it so if you think something else works better I'll switch to that.

**Pull request merged**